toolchain/check/convert.cpp

```diff
@@ -436,30 +445,48 @@ static auto ConvertStructToStruct(Context& context, SemIR::StructType src_type,
   // of the source.
   // TODO: Annotate diagnostics coming from here with the element index.
   CopyOnWriteBlock new_block(sem_ir, literal_elems_id, src_elem_fields.size());
-  for (auto [i, src_field_id, dest_field_id] :
-       llvm::enumerate(src_elem_fields, dest_elem_fields)) {
-    auto src_field = sem_ir.nodes().GetAs<SemIR::StructTypeField>(src_field_id);
+  for (auto [i, dest_field_id] : llvm::enumerate(dest_elem_fields)) {
     auto dest_field =
         sem_ir.nodes().GetAs<SemIR::StructTypeField>(dest_field_id);
-    if (src_field.name_id != dest_field.name_id) {
-      CARBON_DIAGNOSTIC(
-          StructInitFieldNameMismatch, Error,
-          "Mismatched names for field {0} in struct initialization: "
-          "source has field name `{1}`, destination has field name `{2}`.",
-          size_t, llvm::StringRef, llvm::StringRef);
-      context.emitter().Emit(value.parse_node(), StructInitFieldNameMismatch,
-                             i + 1, sem_ir.strings().Get(src_field.name_id),
-                             sem_ir.strings().Get(dest_field.name_id));
-      return SemIR::NodeId::BuiltinError;
+
+    // Find the matching source field.
+    auto src_field_index = i;
+    if (src_type.fields_id != dest_type.fields_id) {
+      auto src_field_it = src_field_indexes.find(dest_field.name_id);
+      if (src_field_it == src_field_indexes.end()) {
+        CARBON_DIAGNOSTIC(
+            StructInitMissingFieldInLiteral, Error,
+            "Missing value for field `{0}` in struct initialization.",
+            llvm::StringRef);
+        CARBON_DIAGNOSTIC(StructInitMissingFieldInConversion, Error,
+                          "Cannot convert from struct type `{0}` to `{1}`: "
+                          "missing field `{2}` in source type.",
+                          std::string, std::string, llvm::StringRef);
```

Comment:
These definitions seem like they should be in the respective scope that uses them, rather than outside. It'd leave less to read over to make sure emit arguments match.

Reply:
Done.